### ISSUE TITLE
[SR-12303] Swift Package Manager package stopped working after Xcode 11.3, Xcode 12, etc.

### ISSUE THREAD

**Issue by swift-ci** [apple/swift-package-manager#4581]

**Comment by tobischw** [comment]
This bug still appears to be around in Xcode 14 (14A309). Ridiculous.

**Comment by neonichu** [comment]
Trying to summarize this thread since it is unclear to me what "this bug" means since there a multiple problems mentioned:
- there's the original issue (FB7608638) - duplicated symbols, fixed in 12.5 
- an issue with packages in Playgrounds (FB7841039) which might be related and is unresolved
- there's the follow-up issue (FB9084743) - `No such module 'RxSwift'`, which continued to affect certain projects until 13.0

Which of these issues are we talking about here as occurring in 14A309 and if you're not talking about FB7841039, is there an example project I could take a look at?

**Comment by tobischw** [comment]
I believe it may still be related to your last bullet point - FB9084743. Trying to use RxSwift within a test target yields `No such module 'RxCocoaRuntime'` - granted, I am not sure if this is an issue with RxSwift, but this seems relevant: https://github.com/ReactiveX/RxSwift/issues/2057

I'll try and get an example project for you. Thanks for looking into this.

**Comment by TLizer** [comment]
@tobischw I just stumbled on similar issue: building unit tests yielded `missing module XYZ`, when I linked missing module to test target then another showed up etc...
It happened that issue lied in `TEST_HOST` build setting which was incorrect.
For some reason xcode creates path to test host as:
```
TEST_HOST = "$(BUILT_PRODUCTS_DIR)/<YourProductName>.app/$(BUNDLE_EXECUTABLE_FOLDER_PATH)/<YourProductName>
```
After checking resolved value for TEST_HOST it was:
```
"$(BUILT_PRODUCTS_DIR)/<YourProductName>.app//<YourProductName>
```
Which means `$(BUNDLE_EXECUTABLE_FOLDER_PATH)` is empty string or non existing.
Correcting this build setting manually to remove `$(BUNDLE_EXECUTABLE_FOLDER_PATH)` fixed issue for me and now tests are compiling and running on pure SPM project :)

**Comment by minguking** [comment]
if there's anyone still suffering from this error by using RxSwift as SPM, this was the solution in my case, very similar to the solution right above me, but slight difference.
try to fix `TEST_HOST`
from 
```TEST_HOST = "$(BUILT_PRODUCTS_DIR)/<ProductName>.app/$(BUNDLE_EXECUTABLE_FOLDER_PATH)/<ProductName>```
To
```TEST_HOST = "$(BUILT_PRODUCTS_DIR)/<ProductName>.app/<ProductName>```
so, just remove the `/$(BUNDLE_EXECUTABLE_FOLDER_PATH)` part. that's it.

**Comment by ncreated** [comment]
👋 We stepped across the same problem in [`Datadog SDK`](https://github.com/DataDog/dd-sdk-ios). Our `Datadog` (Swift) module depends on `_Datadog_Private` (Objective-c) one. Even though the `modulemap` for both is generated properly, the "SwiftEmitModule" build step is unable to find `_Datadog_Private.modulemap` for the app tests target **if the project was created with Xcode 14.x**.

Root cause seems to be the TEST_HOST path being wrong as @TLizer explains above. We explained more in [DataDog/dd-sdk-ios#1085](https://github.com/DataDog/dd-sdk-ios/issues/1085) - we're happy to give more details 🐶💜🍎, although the reproduction steps are quite straightforward.

It would be great to know if `FB9084743` was classified as an SPM / Xcode bug 🐞 or if vendors like us should work on workarounds on their side.

**Comment by neonichu** [comment]
As far as I can see, FB9084743 is considered to be fixed and the `TEST_HOST` issue may not be tracked in FBs at all, yet.

**Comment by ncreated** [comment]
Thanks for reply @neonichu. I just filled `FB11875280` to report the `TEST_HOST` problem ([rdar://FB11875280](https://openradar.appspot.com/FB11875280)).

**Comment by jsryudev** [comment]
This issue still exist in Xcode 14.3 RC2 (14E222b).
To avoid this issue, we need to use a solution from below.

- [TLizer's solution](https://github.com/apple/swift-package-manager/issues/4581#issuecomment-1333577358)
- [minguking's solution](https://github.com/apple/swift-package-manager/issues/4581#issuecomment-1336483409)

**Comment by vani2** [comment]
The solutions above don't work for me.

**Comment by neonichu** [comment]
Would suggest filing feedback reports if this is still happening. FB11875280 is also still open as far as I can see.

**Comment by tommyming** [comment]
The same problem still exists, using Xcode 15 (15A240d), used the workaround solution below, trying to import `RxSwift` and `RxTest` to the unit test target. 

Tried to remove the mentioned packages above, but the same error occurred with different package names.

Please feel free to correct me if there is anything wrong, and I am glad to help, thanks.

> This issue still exist in Xcode 14.3 (14E222b). To avoid this issue, we need to use a solution from below.
> 
> * [TLizer's solution](https://github.com/apple/swift-package-manager/issues/4581#issuecomment-1333577358)
> * [minguking's solution](https://github.com/apple/swift-package-manager/issues/4581#issuecomment-1336483409)

Below is the build result of the test target, suppose I don't need those packages, but Xcode still detects them as required modules.
![Screenshot 2023-09-21 at 12 43 40 PM](https://github.com/apple/swift-package-manager/assets/9570653/1a104e7c-aeb3-46f9-9144-7b85a08d3d95)

**Comment by neonichu** [comment]
Same thing I said in August still applies:

> Would suggest filing feedback reports if this is still happening. FB11875280 is also still open as far as I can see.

**Comment by tommyming** [comment]
Will do that immediately, thanks for the reply!
If there are any updates please let me know also.

**Comment by MaxDesiatov** [comment]
@skv6631 is the issue reproducible for you when building on the command line with `swift build`?

**Comment by alex-pan-invos** [comment]
For anyone still hitting this **with RxSwift specifically** (cc @tommyming @vani2 @tobischw @jsryudev @minguking @Mordil), we put together a fix that has held up on Xcode 26/27 + the arm64 simulator: **[mdata-group/rxswift-dynamic-xcframework](https://github.com/mdata-group/rxswift-dynamic-xcframework)**.

**Full disclosure:** it's our repo, but it does **not** modify RxSwift. It clones the official `ReactiveX/RxSwift` at a pinned tag and repackages the unmodified source as prebuilt **dynamic** `.xcframework`s consumed as SwiftPM `binaryTarget`s (MIT, upstream `LICENSE` bundled). The whole `build.sh` is in the repo and reproducible.

**Why it fixes the two RxSwift-flavoured problems in this thread:**

- **The `getSuperclassMetadata` runtime crash** (the original SR-12303): SwiftPM builds RxSwift *static*, so `RxTest`/`RxBlocking` drag their own full copy of RxSwift into the test bundle while the app embeds another — two RxSwift metadata identities in one process, which the arm64 simulator no longer tolerates. Here `RxSwift.framework` is the single home of the code and `RxCocoa`/`RxRelay`/`RxTest`/`RxBlocking` `@rpath`-link *that same* framework, so host + xctest bundle share **exactly one** copy → one identity → no crash. (`.dynamic` SwiftPM products / `-all_load` / mergeable libraries don't fix this — each still embeds its own copy.)
- **`No such module 'RxSwift'` / `'RxCocoaRuntime'`** (FB9084743 and friends): as a prebuilt binary target there's no per-target source rebuild, so the "swiftmodule not copied into `BUILT_PRODUCTS_DIR`" race can't happen, and `RxCocoaRuntime` is compiled into `RxCocoa.framework` so `import RxCocoa` resolves cleanly.

**Scope, to be honest about it:** this only addresses the **RxSwift** manifestations. It does *not* fix the generic SwiftPM `TEST_HOST` build-setting bug, and it won't help non-RxSwift cases (e.g. the DataDog `_Datadog_Private` reports above) — those need their own dynamic-framework packaging.

Usage (module names are identical, existing `import RxSwift` needs no changes):

```swift
.package(url: "https://github.com/mdata-group/rxswift-dynamic-xcframework", exact: "6.7.0"),
// app target: RxSwift / RxCocoa / RxRelay  (Embed & Sign)
// test target: RxTest / RxBlocking         (Do Not Embed — host already provides RxSwift)
```

`build.sh <version>` regenerates everything for any upstream RxSwift tag. Hope it saves someone the days we lost to this.

**Comment by bkhouri** [comment]
@alex-pan-invos @skv6631  @tommyming : I was able to build https://github.com/ReactiveX/RxSwift project successfully when using the Swift 6.4.x nightly developer snapshot from July 17, 2026. 

Reading the GitHub issue details, the reproduction steps are not available.   Does that mean this issue is fixed?  If not, could you indicate reproduction steps with a minimal reproducer?
